### DISCUSSION TITLE
fn camera control_control_system

### DISCUSSION THREAD

**Discussion by SensorEvolve** [bevyengine/bevy#14503] [Q&A]
I get errors below at //Apply the transformations to the camera, any help would be deeply appreciated.

`fn camera_control_system(
    mut mouse_motion_events: EventReader<MouseMotion>,
    mut mouse_wheel_events: EventReader<MouseWheel>,
    mut query: Query<&mut Transform, With<Camera>>,
) {
    let mut rotation_delta = Vec2::ZERO;
    let mut zoom_delta = 0.0;

    // Process mouse motion events
    for event in mouse_motion_events.read() {
        rotation_delta += event.delta;
    }

    // Process mouse wheel events
    for event in mouse_wheel_events.read() {
        zoom_delta += event.y;
    }

    // Apply the transformations to the camera
    if let Ok(mut transform) = query.single_mut() {
After this line i get this errors and i have no clue why? ■■ this expression has type `Mut<'_, bevy::prelude::Transform>`

        // Rotate the camera around the Y axis (horizontal movement)
        let yaw_rotation = Quat::from_axis_angle(Vec3::Y, -rotation_delta.x * 0.005);
        // Rotate the camera around the local X axis (vertical movement)
        let pitch_rotation = Quat::from_axis_angle(transform.rotation * Vec3::X, -rotation_delta.y * 0.005);

        transform.rotation = yaw_rotation * pitch_rotation * transform.rotation;

        // Move the camera forward/backward based on zoom
        transform.translation += transform.forward() * zoom_delta * 10.0;
    }
}
`

**Comment by piedoom** [comment]
It looks like you have this in your parameters:

```rs
fn camera_control_system(
    mut mouse_motion_events: EventReader,
    mut mouse_wheel_events: EventReader,
    mut query: Query<&mut Transform, With>,
) {
```

`EventReader` and `With` require a generic. In this case, I think you'd want the following ([see](https://github.com/bevyengine/bevy/blob/59a33e6e1477f717da0e72ebcc84aa328e2f5c34/examples/input/mouse_input_events.rs#L19)): 

```rs
fn camera_control_system(
    mut mouse_motion_events: EventReader<MouseMotion>,
    mut mouse_wheel_events: EventReader<MouseWheel>,
    mut query: Query<&mut Transform>,
) {
```

You can drop the `With` here. It's useful when you need to ensure a component exists on an entity, but you don't care about the data. In this case, it looks like you just need a transform. When you query for a component, it will ensure the entities returned have that component, so doing something like `Query<&Transform, With<Transform>>`isn't necessary.

**Reply by SensorEvolve** [reply]
Thank you for your effort. I still got the same errors...

if let Ok(mut transform) = query.single_mut() {    

error[E0308]: mismatched types                                                                                          
   14    --> src/main.rs:239:12                                                                                               
   13     |                                                                                                                   
   12 239 |     if let Ok(mut transform) = query.single_mut() {                                                               
   11     |            ^^^^^^^^^^^^^^^^^   ------------------ this expression has type `Mut<'_, bevy::prelude::Transform>`    
   10     |            |                                                                                                      
    9     |            expected `Mut<'_, Transform>`, found `Result<_, _>`                                                    
    8     |                                                                                                                   
    7     = note: expected struct `Mut<'_, bevy::prelude::Transform>`                                                         
    6                  found enum `Result<_, _>`                                                                              
    5                                                                                                                         
    4 For more information about this error, try `rustc --explain E0308`.                                                     
    3 error: could not compile `globe_real_earth` (bin "globe_real_earth") due to 1 previous error

**Reply by piedoom** [reply]
`single_mut` will either return the query or panic, without any errors or results. You might be looking for `get_single_mut`

**Reply by SensorEvolve** [reply]
Thats it! You solved it!! Thanks m8!!
